### PR TITLE
extmod/modurandom: Add "urandom" module.

### PR DESCRIPTION
**Pull request by pfalcon**
Seedable and reproducible pseudo-random number generator. Implemented
functions are getrandbits(n) (n <= 30) and seed().

The algorithm used is Yasmarang by Ilya Levin:
http://www.literatecode.com/yasmarang

### CODE REVIEW

**Comment by pfalcon** [comment]
I wanted to add pseuso-random number generator for a long time. One issue is adding a whole new module for a single function (this patch adds ~440 bytes on x86_64, good part of that code is module boilerplating), so I had an idea to add it to "sys" instead. I went for a separate module after all, but after coding it, I got an idea how to get best of both solution: add .getrandbits() to sys, but make urandom an alias to sys. That would be my preferred solution now, but submitting the original code for review for now.

**Comment by dpgeorge** [comment]
Yes, this is definitely a good addition!

On stmhal it costs 304 bytes as it is at the moment.  Putting getrandbits and seed functions in sys module saves 36 bytes from that (that's the cost of the `__name__` entry, a dict, and a modules object, 2 + 4 + 3 words).  I don't think that saving is worth it for the unconventionality of having urandom point to sys (but if we do like that pattern, why not just put everything in one grand module? :)

I guess then you are going for option 2 of #965.  That's a good choice I'd say.  A crypto-secure PRNG can be supplied by ssl.RAND_bytes if ever needed.

File: extmod/modurandom.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+   d ^= (pad<<31) + (pad>>1);
+   dat ^= (char) pad ^ (d>>8) ^ 1;
+
+   return (pad^(d<<5)^(pad>>18)^(dat<<1));
+}  /* yasmarang */
+
+// End of Yasmarang
+
+STATIC mp_obj_t mod_urandom_getrandbits(mp_obj_t num_in) {
+    int n = MP_OBJ_SMALL_INT_VALUE(num_in);
```
Comment (dpgeorge) [inline extmod/modurandom.c]: Why not mp_obj_get_int?  Using that saves potential issues where people make small numbers from a bignum (eg 1234567890123456 & 0xff).  That's not common, but if you hit that issue then it's really strange and hard to work out what is happening.

File: extmod/modurandom.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    if (n > 30) {
+        nlr_raise(mp_obj_new_exception(&mp_type_ValueError));
+    }
+    return MP_OBJ_NEW_SMALL_INT(yasmarang() & ((1 << n) - 1));
```
Comment (dpgeorge) [inline extmod/modurandom.c]: Why not allow the full 32 bits and construct with mp_obj_new_int_from_uint?

File: extmod/modurandom.c [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(mod_urandom_getrandbits_obj, mod_urandom_getrandbits);
+
+STATIC mp_obj_t mod_urandom_seed(mp_obj_t seed_in) {
+    mp_uint_t seed = mp_obj_get_int(seed_in);
```
Comment (dpgeorge) [inline extmod/modurandom.c]: Maybe mp_obj_get_int_truncated would be better, then you could pass along a number with the high bit set.

File: extmod/modurandom.c [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+#if MICROPY_PY_URANDOM
+
+// Yasmarang random number generator
+// by Ilya Levin
+// http://www.literatecode.com/yasmarang
+// Public Domain
+
+STATIC uint32_t pad = 0xeda4baba, n = 69, d = 233;
+STATIC uint8_t dat = 0;
```
Comment (dpgeorge) [inline extmod/modurandom.c]: If STATIC is ever defined to nothing (for debugging purposes) won't these small names possibly clash with others?

**Comment by pfalcon** [comment]
> Putting getrandbits and seed functions in sys module saves 36 bytes

On x86_64, it saves 96 bytes, but I actually expected a bit more.

> (but if we do like that pattern, why not just put everything in one grand module? :)

Everything in one module is perhaps too much, but using that pattern when adding single-function-in-a-module is something to keep in mind.

File: extmod/modurandom.c (hunk 1 above)
Comment (pfalcon) [inline extmod/modurandom.c]: Fixed, as well as other int handling cases.

File: extmod/modurandom.c (hunk 4 above)
Comment (pfalcon) [inline extmod/modurandom.c]: I didn't want to change names, to ease comparing with the upstream code. Now prefixed with "yasmarang_".

**Comment by pfalcon** [comment]
Merged with changes described.

**Pull request closed without merging**